bson/_cbsonmodule.c

```diff
@@ -2797,6 +2807,103 @@ static PyObject* _cbson_decode_all(PyObject* self, PyObject* args) {
     return result;
 }
 
+
+static PyObject* _cbson_array_of_documents_to_buffer(PyObject* self, PyObject* args) {
+    uint32_t size;
+    uint32_t value_length;
+    uint32_t position = 0;
+    int32_t initial_position;
+    buffer_t buffer;
+    const char* string;
+    PyObject* arr;
+    PyObject* result = NULL;
+    Py_buffer view = {0};
+
+    if (!PyArg_ParseTuple(args, "O", &arr)) {
+        return NULL;
+    }
+
+    if (!_get_buffer(arr, &view)) {
+        return NULL;
+    }
+
+    buffer = pymongo_buffer_new();
+    if (!buffer) {
+        goto fail;
+    }
+
+    string = (char*)view.buf;
+
+    memcpy(&size, string, 4);
+    size = BSON_UINT32_FROM_LE(size);
+    /* save space for length */
+    if (pymongo_buffer_save_space(buffer, size) == -1) {
+        goto fail;
+    }
+    initial_position = pymongo_buffer_get_position(buffer);
+
+    if (view.len < size || size > BSON_MAX_SIZE) {
+        PyObject* InvalidBSON = _error("InvalidBSON");
+        if (InvalidBSON) {
+            PyErr_SetString(InvalidBSON, "objsize too large");
+            Py_DECREF(InvalidBSON);
+        }
+        goto done;
+    }
+
+    position += 4;
+    while (position < size - 1) {
+        // Just skip the keys.
+        position = position + strlen(string + position) + 1;
+
+        memcpy(&value_length, string + position, 4);
```

Comment:
Need to bounds check position before calling memcpy here or else we could read out of bounds memory.

Reply:
Done